### PR TITLE
FUSETOOLS2-2256: Improve project ESLint configuration to leverage TS/…

### PR DESCRIPTION
**Pull request by pospisilf**
…JS best practices

# Pull Request informations

## Rebase & Merge default requirements

1. Green build for main branch
2. Wait 24 hours after PR creation
3. Green job for PR
4. Approved PR

## PR labels default process

- READY_FOR_REVIEW  &rightarrow;  REVIEW_DONE  &rightarrow;  READY_FOR_MERGE

## Tests

- [ ] Are there **Unit tests**?
- [ ] Are there **Integration tests**?
- [ ] Do we need a new **UI test**?

## PR workflow progress

1. [ ] Tagged with relevant **PR labels**
2. [ ] Green **job for PR**
3. [ ] PR was created more than **24 hours ago** or **All committers approved** it
4. [ ] Green **main** branch build

### CODE REVIEW

**Comment by pospisilf** [comment]
Ready for review in this case means only review of configuration. When configuration will be consulted and approved, I will fix all new errors. @apupier @djelinek

File: .eslintrc.json [hunk 1]
```diff
@@ -1,43 +1,67 @@
...
+        "@typescript-eslint/no-unsafe-argument": "warn",
+        "@typescript-eslint/no-unsafe-assignment": "warn",
+        "@typescript-eslint/no-unsafe-member-access": "warn",
+		"@typescript-eslint/no-unsafe-return": "warn",
+        "@typescript-eslint/no-base-to-string": "warn",
+        "@typescript-eslint/no-unsafe-call": "warn",
+        "@typescript-eslint/restrict-plus-operands": "warn",
+        "@typescript-eslint/no-unnecessary-type-assertion": "warn",
+        "@typescript-eslint/restrict-template-expressions": "warn",
+		"@typescript-eslint/no-floating-promises": "warn"
```
Comment (apupier) [inline .eslintrc.json:55]: Why do we need to list all of them, aren't they mostly inherited from default configuration?

File: .eslintrc.json [hunk 2]
```diff
@@ -1,43 +1,67 @@
...
-			"error",
-			{
-				"devDependencies": [
-					"**/test/**",
-					"**/ui-test/**"
-				],
-				"optionalDependencies": false,
-				"peerDependencies": false
-			}
-		]
```
Comment (apupier) [inline .eslintrc.json:41]: this o,e is very important to keep as it was gfollowing an ugly bug were the binary produced was not the one that was used by automated tests

File: tsconfig.json [hunk 3]
```diff
@@ -1,18 +1,26 @@
...
-		"resolveJsonModule": true
+		"lib": [
+			"ES2022"
+		],
+		"sourceMap": true,
+		"rootDir": "src",
+		"skipLibCheck": true,
+		"resolveJsonModule": true,
+		"strict": true, // true causes like million errors here
+		"forceConsistentCasingInFileNames": true
```
Comment (apupier) [inline tsconfig.json:14]: When modifying that, you will need to check manually that it iss till working at runtime when installing the built vsix. there were issues with that in the past

File: .eslintrc.json (hunk 1 above)
Comment (pospisilf) [reply .eslintrc.json:55]: I wanted to have bigger control on settings between error and warning only. I don't think all of this things must be necessary inherited from default configuration - they're are errors by default. If we decide to set them as error, they can be removed them from list.

**Comment by pospisilf** [comment]
We (with @djelinek) decided to strip down some rules for first iteration. The amount of necessary work to fulfill all set goals at beginning was too big. Goal of first iteration is to match configuration with DAP.

File: src/commands/AbstractNewCamelRouteCommand.ts [hunk 4]
```diff
@@ -90,16 +92,28 @@ export abstract class AbstractNewCamelRouteCommand {
...
 		if (name.includes('.')) {
 			return 'Please provide a name without the extension.';
 		}
+
+		if (!this.camelDSL) {
+			return 'TODO: handle situation'; // camelDSL can't be undefined
+		}
+
+		if (!this.workspaceFolder) {
+			return 'TODO: handle situation';
```
[the remark this answers is not in the document]
Comment (pospisilf) [reply src/commands/AbstractNewCamelRouteCommand.ts:105]: This code is not mine so I left TODOs in some places to open discussion with ENG to make decision how to handle this situations. @apupier @djelinek

File: src/commands/NewCamelKameletCommand.ts [hunk 5]
```diff
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-'use strict'
+'use strict';
```
Comment (djelinek) [inline src/commands/NewCamelKameletCommand.ts:17]: question:

- I am wondering, is this usega of `use strict` on the beginning of files necessary if we have `strict` option enabled by default for config? (I mean in files which are covered by eslint during compilation...)

Comment (pospisilf) [reply src/commands/NewCamelKameletCommand.ts:17]: @djelinek In a fact it's not necessary to keep it here. If we delete it here, nothing will change. Found out that it's good to keep it here for clarity - so when anyone reads the code, immediately  knows that this code is running in strict mode. Up to your decision if we want to keep it or remove.

Comment (djelinek) [reply src/commands/NewCamelKameletCommand.ts:17]: ok ok, I don't have strong opinion here, I was just curious. Thanks 🙂

File: src/requirements/requirements.ts [hunk 6]
```diff
@@ -5,14 +5,15 @@
 import { Uri, env } from 'vscode';
 import * as path from 'path';
 import * as fse from 'fs-extra';
-import * as expandHomeDir from 'expand-home-dir';
 import { Commands } from './commands';
 import { checkJavaPreferences } from './settings';
 import { findJavaHomes, getJavaVersion, JavaRuntime } from './findJavaRuntimes';
 
+const expandHomeDir = require('expand-home-dir');
```
Comment (djelinek) [inline src/requirements/requirements.ts:12]: why is this changed to `require`?

File: src/requirements/requirements.ts [hunk 7]
```diff
@@ -97,7 +103,7 @@ export function parseMajorVersion(version: string): number {
     return javaVersion;
 }
 
-function openJDKDownload(reject, cause) {
+function openJDKDownload(reject: any, cause: any) {
```
Comment (djelinek) [inline src/requirements/requirements.ts:106]: `cause` could be a string instead of any?

**Comment by djelinek** [comment]
need to be rebased

File: src/commands/NewCamelKameletCommand.ts [hunk 8]
```diff
@@ -52,22 +53,26 @@ export class NewCamelKameletCommand extends AbstractNewCamelRouteCommand {
 	protected async showInputBoxForFileName(): Promise<string> {
 		return await window.showInputBox({
 			prompt: this.fileNameInputPrompt,
-			placeHolder: this.camelDSL.placeHolder,
+			placeHolder: this.camelDSL?.placeHolder,
 			validateInput: (fileName) => {
 				return this.validateCamelFileName(`${fileName}-${this.kameletType}`);
 			},
-		});
+		}) || '';
```
Comment (apupier) [inline src/commands/NewCamelKameletCommand.ts:60]: Sonarlint is mentioning (not sure what it implies exactly):

```
Prefer using nullish coalescing operator (`??`) instead of a logical or (`||`), as it is a safer operator.sonarlint(typescript:S6606)
```

File: .eslintrc.json [hunk 9]
```diff
@@ -1,32 +1,47 @@
...
 		"@typescript-eslint/no-namespace": "off",
 		"@typescript-eslint/no-explicit-any": "off",
+		"@typescript-eslint/naming-convention": [ 
+			"error",
+			{
+			  "selector": "variable",
+			  "format": ["camelCase", "UPPER_CASE"]
+			}
+		  ],
+		"@typescript-eslint/semi": "warn",
```
Comment (apupier) [inline .eslintrc.json:40]: this rule is marked as deprecated https://typescript-eslint.io/rules/semi/

```
Formatting rules now live in [eslint-stylistic](https://eslint.style/). [@stylistic/ts/semi](https://eslint.style/rules/ts/semi) is the replacement for this rule.
See [Deprecating Formatting Rules](https://typescript-eslint.io/blog/deprecating-formatting-rules) for more information.
```

File: .eslintrc.json [hunk 10]
```diff
@@ -1,32 +1,47 @@
...
+		"curly": "warn",
+		"eqeqeq": "warn",
+		"@typescript-eslint/no-unused-vars": "warn",
+		"no-throw-literal": "warn",
```
Comment (apupier) [inline .eslintrc.json:44]: All of them are required by the "recommended" config and we set to warn to not fail the build currently?

Comment (pospisilf) [reply .eslintrc.json:44]: we want them as warning mainly for local development purpose - it's really not necessary to get errors because of unused variable or missing semicolon, on the other hand it's necessary to watch these rules so we won't accept PR with errors

File: src/commands/NewCamelKameletCommand.ts (hunk 8 above)
Comment (pospisilf) [reply src/commands/NewCamelKameletCommand.ts:60]: not sure as well, will investigate

File: src/requirements/requirements.ts (hunk 6 above)
Comment (pospisilf) [reply src/requirements/requirements.ts:12]: I don't exactly remember but I was getting some error which was solved by this. Also this way is used in readme of exapand-home-dir. https://www.npmjs.com/package/expand-home-dir

**Review by djelinek — APPROVED** [review]

**Review by apupier — COMMENTED** [review]
Is it possible to rework the commit history? There is a lot of "fixing conflict" or "fixing error" without more details. I guess a bunch of the commit can be squashed and some others can be more precise on what they are doing.
Would be nice to include the FUSETOOLS2-2256 on each of the commit too

**Comment by djelinek** [comment]
> Is it possible to rework the commit history? There is a lot of "fixing conflict" or "fixing error" without more details. I guess a bunch of the commit can be squashed and some others can be more precise on what they are doing. Would be nice to include the FUSETOOLS2-2256 on each of the commit too

a plan is to squash into one commit with FUSETOOLS2-2256: ...

**Comment by apupier** [comment]
> > > Is it possible to rework the commit history? There is a lot of "fixing conflict" or "fixing error" without more details. I guess a bunch of the commit can be squashed and some others can be more precise on what they are doing. Would be nice to include the FUSETOOLS2-2256 on each of the commit too
> > 
> > 
> > a plan is to squash into one commit with FUSETOOLS2-2256: ...
> 
> @apupier +1, I wanted to keep first commits with message containing specific rule to see why was specific change done. All the "fixing conflict" stuff is just temporary and all will be squashed into one commit.

Please update the commit history before merging. It is too easy to make a mistake when doing the merge, like deleting slitghtly too much and pressing enter by mistake before readding the missing part

It can also makes sense to keep some of the commits separated. For instance, when one rule is modified and the corresponding fixes in the code.

File: .eslintrc.json
Comment (pospisilf) [inline .eslintrc.json:1]: only one rule is new required https://eslint.style/guide/migration#approach-1-migrate-to-single-plugin

**Comment by djelinek** [comment]
typos in one commit 
image

- imrpove -> improve
- formating -> formatting
